### DISCUSSION TITLE
How to get the Active Selection objects position relative to canvas

### DISCUSSION THREAD

**Discussion by sohailahmad07** [fabricjs/fabric.js#9943] [Q&A]
Hi when i select multiple objects the position becomes relative to object but i want to get the object's position relative to the canvas

**what i am trying to achieve **

i have a canvas and all of the canvas data is store on the backend in a database i save the json of every object as a separate item assign to a user 

Elemenets -> table_name
id
ele_json
user_id
board_id

**Problem**

the problem is when the user selects multiple objects and moves i send an Axios request to my backend to update object JSON in the database and when i refresh (to simulate the user coming back to the board on a later time) the object are no longer there because when i send the data all of the object position were according to the active selection and not canvas

**Comment by asturur** [comment]
The export of an active selection today is backed in the canvas export and is not great.
The canvas does this:

```
  /**
   * Realizes an object's group transformation on it
   * @private
   * @param {FabricObject} [instance] the object to transform (gets mutated)
   * @returns the original values of instance which were changed
   */
  private _realizeGroupTransformOnObject(
    instance: FabricObject
  ): Partial<typeof instance> {
    const { group } = instance;
    if (group && isActiveSelection(group) && this._activeObject === group) {
      const layoutProps = [
        'angle',
        'flipX',
        'flipY',
        LEFT,
        'scaleX',
        'scaleY',
        'skewX',
        'skewY',
        TOP,
      ] as (keyof typeof instance)[];
      const originalValues = pick<typeof instance>(instance, layoutProps);
      addTransformToObject(instance, group.calcOwnMatrix());
      return originalValues;
    } else {
      return {};
    }
  }
```

It saves the current object transform state,
it apply back the group transform
it exports
and it puts back the original values

It is not a good flow and prone to errors, but as of now it is how is done

**Reply by sohailahmad07** [reply]
Hi sir can you show me how to fix it i am not that good at javascript


 i don't how to call the private method this is how i am exporting the objects 


```
axios.post('url', {
resource: objs.toObject([...customer properties here])
})
```

and the on the backend i loop through the object and add each element into the table as it own record

**Reply by asturur** [reply]
If you export them like this it won't work.
You have to capture the current state, keep track of it, modify the objects, reapply the old state.
Is not convenient

```ts
const { group } = object;
    if (group && isActiveSelection(group)) {
      const layoutProps = [
        'angle',
        'flipX',
        'flipY',
        LEFT,
        'scaleX',
        'scaleY',
        'skewX',
        'skewY',
        TOP,
      ];
      // save original values
      const originalValues = pick(object, layoutProps);
      addTransformToObject(object, group.calcOwnMatrix());
      const toObject = object.toObject(); // this should be what you look for
      // put back original values
      Object.assign(object, originalValues);
```

**Reply by sohailahmad07** [reply]
Hi sir i don't have methods like `pick` or `addTransformToObject` Are those your custom method


i know this is a tall ask be can you give me the entire code 

i found the other method but is not that good especially if you have more objects selected  

```
//  this code is trigger on event : object:modified

   let modifiedObjects = objs.map(obj => {
                let newObj = useFabric.util.object.clone(obj);

                if (newObj.group) {
                    newObj.top =
                        newObj.top + newObj.group.top + newObj.group.height / 2;
                    newObj.left =
                        newObj.left +
                        newObj.group.left +
                        newObj.group.width / 2;
                    return newObj.toObject(properties);
                } else {
                    return newObj.toObject(properties);
                }
            });
            
    axios.post('url', {modifiedObjects})        
```

**Reply by asturur** [reply]
those methods are fabric exports.
On what fabric version are you?

**Reply by sohailahmad07** [reply]
HI Sir 

`"fabric": "^5.3.0",`

i would appreciate if you can the code in a bit detail right now the above code i give will create trigger when `object:modified` **EVENT** is trigger 


Thanks in advance

**Reply by asturur** [reply]
Sorry i can't write out all the solution for fabric 5.3, you could have mentioned it when you opened the post, at least i would have used my time in that way.
But if you look at the canvas function toObject and how it handle the active selection, you can just copy the code from there.

**Reply by sohailahmad07** [reply]
ok i can upgrade the fabric to next stable version if the solution is easy there

Should i upgrade to 6x

**Reply by asturur** [reply]
But if you open fabric 5.3 Canvas toObject method the solution is there by reading the code, what is wrong about that?
